### PR TITLE
[Feat] support DFlash

### PR DESCRIPTION
**Pull request by lingebeng**
## Motivation

Add DFlash speculative decoding support to sglang-jax for `Qwen3-8B-DFlash-b16` on TPU. The goal is to improve greedy decoding throughput while preserving target-model accuracy and supporting continuous batching, TP/DP, and overlap scheduling.

## Modifications

- Added DFlash configuration parsing, draft model implementation, and greedy draft/verify/accept pipeline.
- Integrated Draft KV cache with paged FlashAttention and scheduler memory management.
- Added TP and DP attention support with shape-stable JIT and precompilation.
- Fused draft generation, target verification, and Draft KV materialization to reduce dispatch overhead.
- Added speculative overlap using device-side relay buffers.
- Added unit tests
## Accuracy Tests

> evalscope v1.9.0

Mode | Dataset | Accuracy | num
-- | -- | -- | --
No DFlash | GSM8K | 0.9249 | 1220/1319
DFlash | GSM8K | 0.9257 | 1221/1319

#### prefill
image

#### decode

image

### Environment

- Hardware: TPU v6e-4
- Branch: `epic/dflash`
- Commit: `00f5c182a016054505a2574cdf38b8d5d0ec352a`
- Dataset: random
- Input/output length: 1024/4096
- Random range ratio: 1.0
- Number of prompts: `3 × concurrency`
- DFlash block size: 16
- jax:0.8.1

### Results

The table reports average output throughput as `DFlash / baseline (speedup)`.

| Configuration | Concurrency 16 | Concurrency 32 | Concurrency 64 | Concurrency 128 |
| --- | ---: | ---: | ---: | ---: |
| TP4, DP1, no overlap | 7,331 / 1,663 (**4.41×**) | 9,524 / 2,760 (**3.45×**) | 11,421 / 3,837 (**2.98×**) | 10,858 / 3,917 (**2.77×**) |
| TP4, DP2, no overlap | 5,010 / 1,484 (**3.38×**) | 9,090 / 2,467 (**3.69×**) | 10,823 / 3,361 (**3.22×**) | 10,839 / 3,342 (**3.24×**) |
| TP4, DP2, overlap | 7,914 / 1,839 (**4.30×**) | 9,718 / 3,054 (**3.18×**) | 11,873 / 4,159 (**2.85×**) | 11,545 / 4,368 (**2.64×**) |
| TP4, DP1, overlap | 9,739 / 2,627 (**3.71×**) | 11,687 / 4,007 (**2.92×**) | 13,151 / 4,707 (**2.79×**) | 11,978 / 5,314 (**2.25×**) |

image

The best average output throughput was **13,151 tok/s** at concurrency 64 with TP4, DP1, and overlap scheduling. Peak output throughput reached **22,379 tok/s**. The average accepted length remained between **7.08 and 8.17 tokens** per verification step.

#### with dflash server

```bash

#!/usr/bin/env bash
set -e

pkill -9 -f "[s]gl_jax.launch_server" 2>/dev/null || true
sleep 2

export JAX_COMPILATION_CACHE_DIR=/tmp/jax_cache
export HF_HUB_CACHE=/models
export HF_HOME=/models
export PYTHONPATH=python

.venv/bin/python -m sgl_jax.launch_server \
  --model-path /models/Qwen/Qwen3-8B \
  --speculative-algorithm DFLASH \
  --speculative-draft-model-path /models/z-lab/Qwen3-8B-DFlash-b16 \
  --speculative-num-steps 1 \
  --speculative-eagle-topk 1 \
  --tp-size 4 \
  --dp-size 2 \
  --dtype bfloat16 \
  --attention-backend fa \
  --mem-fraction-static 0.65 \
  --page-size 64 \
  --chunked-prefill-size 2048 \
  --max-running-requests 128 \
  --trust-remote-code \
  --disable-radix-cache \
  --grammar-backend none \
  --host 0.0.0.0 \
  --port 30000

# --disable-overlap-schedule
# --disable-precompile

```

#### without dflash server

```bash
#!/usr/bin/env bash
set -e

pkill -9 -f "[s]gl_jax.launch_server" 2>/dev/null || true
sleep 2

export JAX_COMPILATION_CACHE_DIR=/tmp/jax_cache
export HF_HUB_CACHE=/models
export HF_HOME=/models
export PYTHONPATH=python

.venv/bin/python -m sgl_jax.launch_server \
  --model-path /models/Qwen/Qwen3-8B \
  --trust-remote-code \
  --tp-size 4 \
  --dp-size 2 \
  --mem-fraction-static 0.65 \
  --disable-overlap-schedule \
  --chunked-prefill-size 2048 \
  --dtype bfloat16 \
  --max-running-requests 128 \
  --page-size 64 \
  --attention-backend fa \
  --disable-radix-cache \
  --grammar-backend none \
  --host 0.0.0.0 \
  --port 30000
```
#### bench serving

```bash
#!/usr/bin/env bash
set -e

export PYTHONPATH=python

OUT_DIR=${OUT_DIR:-/home/gcpuser/sky_workdir/bench_results/$(date +%Y%m%d_%H%M%S)}
mkdir -p "$OUT_DIR"

for C in 16 32 64 128; do
  N=$((C * 3))
  mkdir -p "$OUT_DIR/bs_$C"
  .venv/bin/python -m sgl_jax.bench_serving \
    --backend sgl-jax \
    --base-url http://127.0.0.1:30000 \
    --model /models/Qwen/Qwen3-8B \
    --tokenizer /models/Qwen/Qwen3-8B \
    --dataset-name random \
    --random-input-len 1024 \
    --random-output-len 4096 \
    --random-range-ratio 1.0 \
    --num-prompts "$N" \
    --max-concurrency "$C" \
    --disable-tqdm \
    --output-file "$OUT_DIR/bs_$C/result.jsonl"
done
```
#### bench accuracy
```bash
uv run evalscope eval \
    --model /models/Qwen/Qwen3-8B \
    --eval-type openai_api \
    --api-url http://127.0.0.1:30000/v1/chat/completions  \
    --datasets  gsm8k \
    --eval-batch-size 16 \
    --generation-config '{"max_tokens": 4096, "temperature": 0.0}' \
    --work-dir /tmp/evalscope_results
```

## Some Conceptual Diagram

image

image

### CODE REVIEW

File: python/sgl_jax/srt/models/dflash.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        expected = self.num_context_features * self.target_hidden_size
+        if target_hidden.shape[-1] != expected:
+            raise ValueError(
+                "DFLASH target hidden feature mismatch: "
+                f"expected last dim {expected}, got {target_hidden.shape[-1]}."
+            )
+        hidden_states, _ = self.fc(target_hidden)
+        return self.hidden_norm(hidden_states)
+
+    # @haifeng kv injection here
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/models/dflash.py]: Pay attention to the comments here.

File: python/sgl_jax/srt/models/dflash.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+            param_dtype=dtype,
+            scope_name="post_attention_layernorm",
+        )
+        self.mlp = Qwen3MLP(
+            hidden_size=hidden_size,
+            intermediate_size=int(config.intermediate_size),
+            mesh=mesh,
+            layer_id=layer_id,
+            dtype=dtype,
+        )
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/models/dflash.py]: Do the MLPs for dflash in the other models here all use the Qwen3 MLP architecture?

File: python/sgl_jax/srt/speculative/draft_extend_fused.py [hunk 3]
```diff
@@ -579,13 +579,26 @@ def fused_draft_extend(
...
+    lens_2d = _reshape_per_dp_rows(lens, dp_size)
     zeros = jnp.zeros_like(lens_2d[:, :1], dtype=jnp.int32)
-    return jnp.concatenate([zeros, jnp.cumsum(lens_2d, axis=1, dtype=jnp.int32)], axis=1).reshape(
+    result = jnp.concatenate([zeros, jnp.cumsum(lens_2d, axis=1, dtype=jnp.int32)], axis=1).reshape(
         (dp_size * (per_dp_bs + 1),)
     )
+    sharding = jax.typeof(lens).sharding
+    if isinstance(sharding, NamedSharding) and not sharding.mesh.empty:
+        result = jax.sharding.reshard(result, sharding)
+    return result
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/speculative/draft_extend_fused.py:602]: Explain why resharding is necessary here.

File: python/sgl_jax/test/speculative/test_dflash_info.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    build_dflash_draft_block,
+    dflash_greedy_verify,
+)
+from sgl_jax.srt.speculative.overlap_utils import (
+    can_merge_spec_non_overlap_prefill,
+    use_legacy_eagle3_non_overlap,
+)
+from sgl_jax.srt.speculative.spec_info import SpeculativeAlgorithm
+
+
```
Comment (JamesBrianD) [inline python/sgl_jax/test/speculative/test_dflash_info.py:18]: The unit test was not included in the test suite.

File: python/sgl_jax/srt/managers/scheduler.py [hunk 5]
```diff
@@ -138,14 +140,33 @@ class GenerationBatchResult:
...
+        return "DFLASH speculative decoding does not support return_logprob yet."
+    sp = req.sampling_params
+    if (
+        getattr(sp, "json_schema", None) is not None
+        or getattr(sp, "regex", None) is not None
+        or getattr(sp, "ebnf", None) is not None
+        or getattr(sp, "structural_tag", None) is not None
+    ):
+        return "DFLASH speculative decoding does not support grammar-constrained decoding yet."
+    return None
```
Comment (cjx0709) [inline python/sgl_jax/srt/managers/scheduler.py:179]: DFLASH verify always uses argmax, but non-greedy requests can pass this check. If not support non-greedy, maybe force reject this situation.

File: python/sgl_jax/srt/speculative/dflash_info.py [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+            v = getattr(self, f, None)
+            if v is not None:
+                setattr(self, f, np.asarray(v, dtype=np.int32))
+
+    def new_tokens_required_next_decode(self, requests, page_size: int) -> int:
+        total = 0
+        block_size = int(self.block_size)
+        for req in requests:
+            cur = int(req.kv_allocated_len)
+            nxt = max(cur, int(req.kv_committed_len) + block_size)
```
Comment (cjx0709) [inline python/sgl_jax/srt/speculative/dflash_info.py]: Overlap reserves 2 * block_size later, but this only accounts for one block. Could this undercount KV pages and cause allocation failures near a page boundary?

File: python/sgl_jax/srt/managers/scheduler.py [hunk 7]
```diff
@@ -138,14 +140,35 @@ class GenerationBatchResult:
...
+        return "DFLASH speculative decoding does not support return_logprob yet."
+    sp = req.sampling_params
+    if (
+        getattr(sp, "json_schema", None) is not None
+        or getattr(sp, "regex", None) is not None
+        or getattr(sp, "ebnf", None) is not None
+        or getattr(sp, "structural_tag", None) is not None
+    ):
+        return "DFLASH speculative decoding does not support grammar-constrained decoding yet."
+    if sp.top_k != 1:
```
Comment (cjx0709) [inline python/sgl_jax/srt/managers/scheduler.py:167]: Could we reject penalty-based requests here too? They skip fused prefill; then overlap decode expects relay state and fails. Without overlap, the penalty is silently ignored. Rejecting frequency/presence/repetition penalties and min_new_tokens for now should be enough.

**Review by cjx0709 — APPROVED** [review]
LGTM
